File: swapNodesInPairs.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        r = head
        pre = None
        while head and head.next:
            # l = r
            # ret = ""
            # while l:
            #     ret += str(l.val)+"-"
            #     l = l.next
            # print(ret)

            if pre == None:
                r = head.next
                head.next = head.next.next
                r.next = head
                pre = head
                head = head.next
            else:
                n = head.next
                head.next = head.next.next
                pre.next = n
                n.next = head
                pre = head
                head = head.next
        return r

l = ListNode(1)
l.next = ListNode(2)
l.next.next = ListNode(3)
l.next.next.next = ListNode(4)
l.next.next.next.next = ListNode(5)

l = Solution().swapPairs(l)
while l:
    print(l.val)
    l = l.next

File: LeeCode/swapNodesInPairs.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init() { self.val = 0; self.next = nil; }
 *     public init(_ val: Int) { self.val = val; self.next = nil; }
 *     public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
 * }
 */
class Solution {
    func swapPairs(_ head: ListNode?) -> ListNode? {
        var temp = head
        var h: ListNode? = head
        var pre: ListNode? = nil
        while temp != nil && temp?.next != nil {
            let tempN = temp?.next?.next
            if pre != nil {
                pre?.next = temp?.next
                temp?.next?.next = temp
                pre = temp
                temp?.next = tempN
                temp = tempN
            } else {
                h = temp?.next
                pre = temp
                h?.next = temp
                temp?.next = tempN
                temp = tempN
            }
        }
        return h
    }
}